Implement a safer `Resource.attempt` which releases acquired resource(s) in case of error

lenguyenthanh opened the pull request:
This is my solution for #3757, 

As I discussed with @armanbilge, the new method requires `MonadCancelThrow` in order to use `Resource.allocatedCase` which handles releasing acquired resource(s) in case of error. This pr also removed implicit `ApplicativeError` and deprecated the current attempt method.

~~This is a draft as it still needs to add tests and address some issues that I'll comment in the pr. But please give me early feedback.~~

lenguyenthanh committed:
Implement a safer Resource.attempt, fix #3757

Which releases acquired resource(s) in case of error and `MonadCancelThrow`
constrain is required in order to do that. This also take away implicit
`ApplicativeError` and deprecated the current attempt method.

lenguyenthanh commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -1462,7 +1473,7 @@ abstract private[effect] class ResourceMonadError[F[_], E]
 
   override def attempt[A](fa: Resource[F, A]): Resource[F, Either[E, A]] =
-    fa.attempt
+    fa.attempt(F)
```
this still used the old `attempt` but We can't use the new one because the error type is not fixed to `Throwable`.

armanbilge replied:
We should mark this instance as deprecated and create a new instance that requires `MonadCancelThrow`.

lenguyenthanh replied:
fixed in https://github.com/typelevel/cats-effect/pull/4128/commits/cb1a293b49af10ac9127ba23fbb69d801465ea79. But now https://github.com/typelevel/cats-effect/blob/cb1a293b49af10ac9127ba23fbb69d801465ea79/kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala#L1300-L1306 complains because it use the deprecated `ResourceMonadError`.

I'm not sure what the best way to fix this. I feel like We can remove it completely. But maybe it broke client code.

armanbilge replied:
We deprecate this one too :)

lenguyenthanh replied:
We deprecate this one but keep the trait hierarchy?

armanbilge replied:
That's a good question. Maybe we should try to deprioritize this instance by moving it to the bottom of the hierarchy. Alternatively, we could just remove the `implicit` modifier.

lenguyenthanh replied:
I kept the hierachy the same but removed implicit modifier and deprecated `catsEffectMonadErrorForResource` function.

armanbilge replied:
I thought about it more and actually I think it's okay to leave the `implicit` there. We don't need to change any prioritization, because we aren't introducing any new instances, and the instance derived from `MonadCancelThrow` (which is required for the fixed `attempt`) is already higher priority than this one.

By leaving it `implicit`, users will get an actionable deprecation instead of non-compiling code.

lenguyenthanh commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -710,19 +712,28 @@ sealed abstract class Resource[F[_], +A] extends Serializable {
   def handleErrorWith[B >: A, E](f: E => Resource[F, B])(
       implicit F: ApplicativeError[F, E]): Resource[F, B] =
-    attempt.flatMap {
+    attempt(F).flatMap {
```
imo, We should add a "safe" variant of `handleErrorWith` as well to avoid inconsistent behaviors between this and `attempt`.

armanbilge replied:
Good catch. You are right, we definitely should do that.

lenguyenthanh replied:
fixed in https://github.com/typelevel/cats-effect/pull/4128/commits/3c2d6af91df66c3d71088c9c6ebad884c06615e2

armanbilge commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -700,7 +700,9 @@ sealed abstract class Resource[F[_], +A] extends Serializable {
 
-  def attempt[E](implicit F: ApplicativeError[F, E]): Resource[F, Either[E, A]] =
+  @deprecated("Use overload with MonadCancelThrow", "3.6.0")
+  def attempt[E](F: ApplicativeError[F, E]): Resource[F, Either[E, A]] = {
+    implicit val x: ApplicativeError[F, E] = F
```
We can do a trick here to automatically fix more callsites without requirng them to recompile: we can pattern match on `F` and if it is `Sync` or `Async`  then we can use the fixed implementation instead.

lenguyenthanh replied:
I'm not sure how to implement your suggestion.

I tried:

```scala
  @deprecated("Use overload with MonadCancelThrow", "3.6.0")
  def attempt[E](F: ApplicativeError[F, E]): Resource[F, Either[E, A]] =
    F match {
      case x: Async[_] =>
        implicit val x: MonadCancelThrow[F] = x
        attempt
      case x: Sync[_] => attempt(x)
        implicit val x: MonadCancelThrow[F] = x
        attempt
      case _ => ???
   }
```

But it doesn't work as `attempt` with `MonadCancelThrow` returns `Resource[F, Either[Throwable, A]]` but not `Resource[F, Either[E, A]]`

armanbilge replied:
You can cast unsafely. If `F` is `Sync` then `E` must be `Throwable`. Also I just realized since `Async <: Sync` we only need to match for `Sync`.

lenguyenthanh replied:
Fixed it https://github.com/typelevel/cats-effect/pull/4128/commits/3b7d0ee83c4ecfeb639b1fa75493ca6b73c1b823

kubukoz commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -710,19 +712,28 @@ sealed abstract class Resource[F[_], +A] extends Serializable {
+
+  def attempt(implicit F: MonadCancelThrow[F]): Resource[F, Either[Throwable, A]] =
+    Resource.applyCase[F, Either[Throwable, A]] {
+      allocatedCase.attempt.map {
```
**question:** shouldn't the allocation itself be interruptible?

like `Resource.applyFull { poll => poll(allocatedCase).attempt.map...`

armanbilge replied:
Good catch, this looks right. Let's make sure to add a test case for this.

lenguyenthanh replied:
I though about using `poll` but didn't know for sure. I added tests and fix this in https://github.com/typelevel/cats-effect/pull/4128/commits/07fa387a2d555f37196b9fdb0a9a7419b20f5129 and https://github.com/typelevel/cats-effect/pull/4128/commits/28394460eadd78805ccead379d9dbe77ef58c6a2

lenguyenthanh committed:
Add test cases for safe attempt

lenguyenthanh committed:
Make allocatedCase interupible

lenguyenthanh committed:
Implement handleErrorWith with MonadCancelThrow

Deprecated handleErrorWith with ApplicativeError

lenguyenthanh committed:
Deprecate ResourceMonadError

lenguyenthanh commented on core/jvm/src/main/scala/cats/effect/metrics/JvmCpuStarvationMetrics.scala:
```diff
@@ -66,7 +66,7 @@ private[effect] object JvmCpuStarvationMetrics {
       }
       .map(_._2)
-      .handleErrorWith[CpuStarvationMetrics, Throwable] { th =>
+      .handleErrorWith[CpuStarvationMetrics] { th =>
```
This change because `handleErrorWith[E, A]` now requires explicit `ApplicativeError`

lenguyenthanh committed:
Best effort to improve deprecated attempt

Fix F is `Sync`, We know that it has to be a `MonadCancelThrow`,
therefore We can use safer attempt and `unsafely` cast the result.

lenguyenthanh committed:
Deprecate catsEffectMonadErrorForResource

Also remove implicit modifier from it.

armanbilge commented on tests/shared/src/test/scala/cats/effect/ResourceSpec.scala:
```diff
@@ -1172,6 +1172,29 @@ class ResourceSpec extends BaseSpec with ScalaCheck with Discipline {
+        .flatMap { ref =>
+          val resource = Resource.make(ref.update(_ + 1))(_ => ref.update(_ + 1))
+          val error = Resource.raiseError[IO, Unit, Throwable](new Exception)
+          (resource *> error).attempt.use { _ => ref.get.map { _ must be_==(2) } }
```
instead of ignoring the value `_` we can check that it's a `Left`.

armanbilge commented on tests/shared/src/test/scala/cats/effect/ResourceSpec.scala:
```diff
@@ -1172,6 +1172,29 @@ class ResourceSpec extends BaseSpec with ScalaCheck with Discipline {
+    }
+
+    "acquire is interruptible" in ticked { implicit ticker =>
+      val sleep = IO.sleep(1.second)
```
Could also be:
```suggestion
      val sleep = IO.never
```

armanbilge commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -700,29 +700,52 @@ sealed abstract class Resource[F[_], +A] extends Serializable {
+  def attempt(implicit F: MonadCancelThrow[F]): Resource[F, Either[Throwable, A]] =
+    Resource.applyFull[F, Either[Throwable, A]] { poll =>
+      poll(allocatedCase).attempt.map {
+        case Left(error) => (error.asLeft[A], _ => ().pure[F])
+        case Right((a, r)) => (a.asRight[Throwable], r)
```
Some shuffling to avoid allocations and improve performance.

```suggestion
        case Right((a, r)) => (a.asRight[Throwable], r)
        case error => (error, _ => F.unit)
```

armanbilge commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -1330,10 +1354,19 @@ abstract private[effect] class ResourceFOInstances1 {
   implicit protected def F: MonadCancel[F, Throwable]
 
+  override def attempt[A](fa: Resource[F, A]): Resource[F, Either[Throwable, A]] =
+    fa.attempt(F)
```
Unless it's necessary for some reason? 🤔 

```suggestion
    fa.attempt
```

armanbilge commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -700,29 +700,52 @@ sealed abstract class Resource[F[_], +A] extends Serializable {
+
+  def handleErrorWith[B >: A](f: Throwable => Resource[F, B])(
+      implicit F: MonadCancelThrow[F]): Resource[F, B] =
+    attempt(F).flatMap {
```
Unless it's necessary for some reason?

```suggestion
    attempt.flatMap {
```

lenguyenthanh committed:
Revert implicit for Resource MonadError instance

MonadCancelThrow instance (which is required for the fixed attempt)
is already higher priority than this one. By leaving it implicit,
users will get an actionable deprecation instead of non-compiling code.

lenguyenthanh committed:
Verify attempt returns Left in case of error

lenguyenthanh committed:
Apply suggestions from code review

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

lenguyenthanh committed:
Fix compilation error

armanbilge commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -1279,6 +1302,7 @@ private[effect] trait ResourceHOInstances3 extends ResourceHOInstances4 {
 }
 
 private[effect] trait ResourceHOInstances4 extends ResourceHOInstances5 {
+  @deprecated("Use MonadCancelThrow instances", "3.6.0")
```
Maybe a message like this can be more helpful to the user?

```suggestion
  @deprecated("Bring an implicit MonadCancelThrow[F] into scope to get the fixed Resource instance", "3.6.0")
```

armanbilge reviewed (commented):
Looking good, one last comment :)

lenguyenthanh committed:
Better deprecated message

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

armanbilge reviewed (approved):
nice work :)
